internal/stats/aggregate.go

```diff
@@ -0,0 +1,279 @@
+	for i := range reflectType.NumField() {
+		field := reflectType.Field(i)
+
+		// Recursively get all embedded tags.
+		if field.Anonymous {
+			extractJSONTags(field.Type, jsonTags)
+		}
+
+		jsonTag := util.JSONFieldName(field)
+
+		if jsonTag != "" {
+			jsonTags[jsonTag] = true
+		}
+	}
+}
+
+// Group together metrics based on their groupByFields.
+func groupTogetherMetrics(metrics []map[string]any, query AggregationQuery) (map[string][]map[string]string, error) {
+	groupedMetricBuckets := make(map[string][]map[string]string)
+
+	for _, metric := range metrics {
+		// Skip grouping this metric if it doesn't contain all groupByFields and the overviewField.
+		if !hasAllNeededFields(metric, query) {
+			continue
+		}
+
+		// Turn all values into strings for aggregation.
+		stringMetric := make(map[string]string, len(metric))
+		for field, value := range metric {
+			stringMetric[field] = fmt.Sprintf("%v", value)
+		}
+
+		var groupFieldValues []string
+		for _, field := range query.GroupByFields {
+			groupFieldValues = append(groupFieldValues, fmt.Sprintf("%v", metric[field]))
+		}
+
+		bucketKey := strings.Join(groupFieldValues, "::")
+		groupedMetricBuckets[bucketKey] = append(groupedMetricBuckets[bucketKey], stringMetric)
+	}
+
+	return groupedMetricBuckets, nil
+}
+
+func hasAllNeededFields(metric map[string]any, query AggregationQuery) bool {
+	for _, field := range query.GroupByFields {
+		_, exists := metric[field]
+		if !exists {
+			return false
+		}
+	}
+
+	_, exists := metric[query.OverviewField]
+	if !exists {
+		return false
+	}
+
+	return true
+}
+
+// Calculate aggregations for a group of metrics.
+func computeGroupAggregation(groupedMetrics []map[string]string, query MetricQuery) (map[string]any, error) {
+	if len(groupedMetrics) == 0 {
+		return nil, nil
+	}
+
+	aggregatedResults, err := extractFields(groupedMetrics[0], query)
+	if err != nil {
+		return nil, fmt.Errorf("Failed to extract group by fields from metric '%v': '%v'.", groupedMetrics[0], err)
+	}
+
+	// Determine if aggregation should be numeric or non-numeric.
+	isNumeric := canAggregateAsNumeric(groupedMetrics[0], query.OverviewField)
+
+	if isNumeric {
+		return aggregateNumeric(groupedMetrics, aggregatedResults, query.OverviewField)
+	} else {
+		return aggregateNonNumeric(groupedMetrics, aggregatedResults)
+	}
+}
+
+func extractFields(metric map[string]string, query MetricQuery) (map[string]any, error) {
+	aggregatedResults := make(map[string]any)
+	groupByValues := make(map[string]string)
+
+	for _, field := range query.GroupByFields {
+		value := metric[field]
+		groupByValues[field] = value
+	}
+
+	aggregatedResults[GROUP_BY] = groupByValues
+	aggregatedResults[OVERVIEW] = query.OverviewField
+
+	return aggregatedResults, nil
+}
+
+func canAggregateAsNumeric(metric map[string]string, overviewField string) bool {
+	aggregateValue := metric[overviewField]
+	_, err := util.StrToFloat(aggregateValue)
```

Comment:
This is a bit dangerous, since somethings can be converted to numbers but are supposed to be treated as strings (e.g. locators).
I made a comment earlier about keeping the value as its true type instead of converting to a string.
That would solve this.